concordium-std/src/state_btree.rs

```diff
@@ -0,0 +1,1481 @@
+        } else {
+            // Unreachable code, since M must be 2 or larger (the minimum degree), a
+            // child node must have at least one sibling.
+            crate::trap();
+        }
+    }
+
+    /// Internal function for getting the highest key in a subtree.
+    /// Assumes the provided `node` is not a leaf and has a child at
+    /// `child_index`.
+    fn get_highest_key(&self, node: &Node<M, K>, child_index: usize) -> K
+    where
+        K: Ord + Serialize,
+        S: HasStateApi, {
+        let mut node = self.get_node(node.children[child_index]);
+        while !node.is_leaf() {
+            let child_node_id = node.children.last().unwrap_abort();
+            node = self.get_node(*child_node_id);
+        }
+        // This does not mutate the node in the end, just the representation in memory
+        // which is freed after the call.
+        node.keys.pop().unwrap_abort()
+    }
+
+    /// Internal function for getting the lowest key in a subtree.
+    /// Assumes the provided `node` is not a leaf and has a child at
+    /// `child_index`.
+    fn get_lowest_key(&self, node: &Node<M, K>, child_index: usize) -> K
+    where
+        K: Ord + Serialize,
+        S: HasStateApi, {
+        let mut node = self.get_node(node.children[child_index]);
+        while !node.is_leaf() {
+            let child_node_id = node.children.first().unwrap_abort();
+            node = self.get_node(*child_node_id);
+        }
+        // This does not mutate the node in the end, just the representation in memory
+        // which is freed after the call.
+        node.keys.swap_remove(0)
+    }
+
+    /// Moving key at `index` from the node to the lower child and then merges
+    /// this child with the content of its larger sibling, deleting the sibling.
+    ///
+    /// Assumes:
+    /// - `parent_node` has children `child` at `index` and `larger_child` at
+    ///   `index + 1`.
+    /// - Both children are at minimum number of keys (`M - 1`).
+    fn merge(
+        &mut self,
+        parent_node: &mut Node<M, K>,
+        index: usize,
+        child: &mut Node<M, K>,
+        mut larger_child: StateRefMut<Node<M, K>, S>,
+    ) where
+        K: Ord + Serialize,
+        S: HasStateApi, {
+        let parent_key = parent_node.keys.remove(index);
+        parent_node.children.remove(index + 1);
+        child.keys.push(parent_key);
+        child.keys.append(&mut larger_child.keys);
+        child.children.append(&mut larger_child.children);
+        self.delete_node(larger_child);
+    }
+
+    /// Internal function for constructing a node. It will increment the next
+    /// node ID and create an entry in the smart contract key-value store.
+    fn create_node<'b>(
+        &mut self,
+        keys: Vec<K>,
+        children: Vec<NodeId>,
+    ) -> (NodeId, StateRefMut<'b, Node<M, K>, S>)
+    where
+        K: Serialize,
+        S: HasStateApi, {
+        let node_id = self.next_node_id.copy_then_increment();
```

Comment:
These node ids are u32, but I don't see them being reset unless you clear the set fully.

That worries me a bit even though it's probably still quite generous.

If we have 5 inserts per second this would run out in less than 30 years :)

I guess using u64 for node ids is more expensive.

Reply:
I could use something like uleb128 if this is a concern, not sure how much it will impact the cost though

Reply:
This was discussed offline, and the conclusion from that was to first have benchmarks comparing `u32` vs `u64` for the NodeId and decide based on the results.

Reply:
The benchmarks showed a small cost increase. Small enough, so the decision was to use `u64` to remove the concern of running out of node IDs.